okio/src/test/java/okio/ReadUtf8LineTest.java

```diff
@@ -194,6 +194,55 @@
     assertEquals("ghi\rjkl", source.readUtf8LineStrict());
   }
 
+  @Test public void readUtf8LineWithLimits() throws IOException {
+    for (int limit : new int[] {
+        5, 6, 7, 10000
+    }) {
+      data.writeUtf8("panda\r\n");
+      assertEquals("panda", source.readUtf8Line(limit));
+      assertTrue(data.exhausted());
+    }
+
+    // newline not within limit
+    for (String input : new String[] {
+        "panda", "panda\n", "panda\r\n",
+    }) {
+      data.writeUtf8(input);
+      assertEquals("p", source.readUtf8Line(1));
+      assertEquals("an", source.readUtf8Line(2));
+      assertEquals("da", source.readUtf8Line(2));
+      assertTrue(data.exhausted());
+    }
+
+    // Trim trailing newlines
```

Comment:
generally I think it’s better to avoid loops in test cases, instead using helper methods. In this case you could imagine a method like so:

```
private void testTrimTrailingNewlines(String input) {
  Buffer data = new Buffer();
  data.writeUtf8(input);
  assertEquals("panda", source.readUtf8Line(5));
  assertTrue(data.exhausted());
}
```
Then your code reads like so:
```
trimTrailingNewlines("panda");
trimTrailingNewlines("panda\n");
trimTrailingNewlines("panda\r\n");
```

Two reasons why I like this structure better. First, if it fails you get a stacktrace that tells you exactly which input failed. With this a failure on `data.exhausted()` won’t tell you which of the inputs was bad. Second, control flow in test cases is risky because test cases don’t have test cases.